src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+use anyhow::format_err;
+use anyhow::{bail, Context};
+use serde::Deserialize;
+use std::time::Duration;
+
+/// Garbage collector.
+pub struct Gc<'a, 'config> {
+    config: &'config Config,
+    global_cache_tracker: &'a mut GlobalCacheTracker,
+    /// A lock on the package cache.
+    ///
+    /// This is important to be held, since we don't want multiple cargos to
+    /// be allowed to write to the cache at the same time, or for others to
+    /// read while we are modifying the cache.
+    #[allow(dead_code)] // Held for drop.
+    lock: CacheLock<'config>,
+}
+
+/// Automatic garbage collection settings from the `gc.auto` config table.
+///
+/// NOTE: Not all of these options may get stabilized. Some of them are very
+/// low-level details, and may not be something typical users need.
+///
+/// If any of these options are `None`, the built-in default is used.
+#[derive(Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+struct AutoConfig {
+    /// The maximum frequency that automatic garbage collection happens.
+    frequency: Option<String>,
+    /// Anything older than this duration will be deleted in the source cache.
+    max_src_age: Option<String>,
+    /// Anything older than this duration will be deleted in the compressed crate cache.
+    max_crate_age: Option<String>,
+    /// Any index older than this duration will be deleted from the index cache.
+    max_index_age: Option<String>,
+    /// Any git checkout older than this duration will be deleted from the checkout cache.
+    max_git_co_age: Option<String>,
+    /// Any git clone older than this duration will be deleted from the git cache.
+    max_git_db_age: Option<String>,
+}
+
+/// Options to use for garbage collection.
+#[derive(Clone, Debug, Default)]
+pub struct GcOpts {
+    /// The `--max-src-age` CLI option.
+    pub max_src_age: Option<Duration>,
+    // The `--max-crate-age` CLI option.
+    pub max_crate_age: Option<Duration>,
+    /// The `--max-index-age` CLI option.
+    pub max_index_age: Option<Duration>,
+    /// The `--max-git-co-age` CLI option.
+    pub max_git_co_age: Option<Duration>,
+    /// The `--max-git-db-age` CLI option.
+    pub max_git_db_age: Option<Duration>,
+    /// The `--max-src-size` CLI option.
+    pub max_src_size: Option<u64>,
+    /// The `--max-crate-size` CLI option.
+    pub max_crate_size: Option<u64>,
+    /// The `--max-git-size` CLI option.
+    pub max_git_size: Option<u64>,
+    /// The `--max-download-size` CLI option.
+    pub max_download_size: Option<u64>,
+
+    /// The `--max-target-age` CLI option (UNIMPLEMENTED).
+    pub max_target_age: Option<Duration>,
+    /// The `--max-shared-target-age CLI option (UNIMPLEMENTED).
+    pub max_shared_target_age: Option<Duration>,
+    /// The `--max-target-size` CLI option  (UNIMPLEMENTED).
+    pub max_target_size: Option<u64>,
+    /// The `--max-shared-target-size` CLI option (UNIMPLEMENTED).
+    pub max_shared_target_size: Option<u64>,
```

Comment:
The more I see of these, the more uncomfortable I get.  For how big of a PR this is, I don't think we should be trying to also demonstrate other things ,creating a lot of dead code baggage.